### ISSUE TITLE
sqlfluff-templater-dbt==2.1.1 not compatible with dbt-core==1.5.2

### ISSUE THREAD

**Issue by MarcBook** [sqlfluff/sqlfluff#4930]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Executing sqlfluff lint doesn't work.
The issue is most likely caused by this commit in dbt-core.
https://github.com/dbt-labs/dbt-core/commit/ca73a2aa15aba760abd4503dab7a16d6023c389f

### Expected Behaviour

sqlfluff lint should work

### Observed Behaviour

sqlfluff lint isn't working.

```
Traceback (most recent call last):
  File "/usr/local/bin/sqlfluff", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/sqlfluff/cli/commands.py", line 605, in lint
    result = lnt.lint_paths(
  File "/usr/local/lib/python3.10/site-packages/sqlfluff/core/linter/linter.py", line 1206, in lint_paths
    for i, linted_file in enumerate(runner.run(expanded_paths, fix), start=1):
  File "/usr/local/lib/python3.10/site-packages/sqlfluff/core/linter/runner.py", line 106, in run
    for fname, partial in self.iter_partials(fnames, fix=fix):
  File "/usr/local/lib/python3.10/site-packages/sqlfluff/core/linter/runner.py", line 59, in iter_partials
    for fname, rendered in self.iter_rendered(fnames):
  File "/usr/local/lib/python3.10/site-packages/sqlfluff/core/linter/runner.py", line 46, in iter_rendered
    yield fname, self.linter.render_file(fname, self.config)
  File "/usr/local/lib/python3.10/site-packages/sqlfluff/core/linter/linter.py", line 879, in render_file
    return self.render_string(raw_file, fname, config, encoding)
  File "/usr/local/lib/python3.10/site-packages/sqlfluff/core/linter/linter.py", line 850, in render_string
    templated_file, templater_violations = self.templater.process(
  File "/usr/local/lib/python3.10/site-packages/sqlfluff/core/templaters/base.py", line 53, in _wrapped
    return func(self, in_str=in_str, fname=fname, config=config, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/sqlfluff_templater_dbt/templater.py", line 351, in process
    processed_result = self._unsafe_process(fname_absolute_path, in_str, config)
  File "/usr/local/lib/python3.10/site-packages/sqlfluff_templater_dbt/templater.py", line 467, in _unsafe_process
    node = self._find_node(fname, config)
  File "/usr/local/lib/python3.10/site-packages/sqlfluff_templater_dbt/templater.py", line 404, in _find_node
    results = [self.dbt_manifest.expect(uid) for uid in selected]
  File "/usr/local/lib/python3.10/site-packages/sqlfluff_templater_dbt/templater.py", line 404, in <listcomp>
    results = [self.dbt_manifest.expect(uid) for uid in selected]
  File "/usr/local/lib/python3.10/site-packages/dbt/graph/selector_methods.py", line 329, in search
    root = Path(cv_project_root.get())
LookupError: <ContextVar name='project_root' at 0xffffaf887010>
```

### How to reproduce

Install the following packages `pip install sqlfluff==2.1.1 sqlfluff-templater-dbt==2.1.1 dbt-bigquery==1.5.2 dbt-core==1.5.2`
Run `sqlfluff lint` inside your dbt project

Python version: 3.10.11

### Dialect

dialect = bigquery
templater = dbt

### Version

2.1.1

### Configuration

**.sqlfluff:**
```
[sqlfluff]
dialect = bigquery
templater = dbt
large_file_skip_char_limit=0
large_file_skip_byte_limit=0
exclude_rules = L016, L029, L034
max_line_length = 120

[sqlfluff:indentation]
allow_implicit_indents = True

[sqlfluff:templater:dbt]
project_dir = ./
profiles_dir = ./
target = dev

[sqlfluff:templater:dbt:context]
start_date = 2023-01-01

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:rules:capitalisation.keywords]
# Keywords
capitalisation_policy = upper
ignore_words = FIRST, SECOND # words used as column name

[sqlfluff:rules:layout.long_lines]
ignore_comment_lines = True
ignore_comment_clauses = True

[sqlfluff:rules:capitalisation.functions]
# Function names
extended_capitalisation_policy = upper

[sqlfluff:rules:capitalisation.literals]
# Null & Boolean Literals
capitalisation_policy = upper
```

**.sqlfluffignore**:
```
target/
# dbt <1.0.0
dbt_modules/
# dbt >=1.0.0
dbt_packages/
macros/
env/
example/
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by robsicurelli** [comment]
having the same issue. easiest thing for me to do is downgrade dbt-core to 1.5.1

**Comment by hiro-o918** [comment]
I also found that parallelization of sqlfluff causes another error for dbt-core 1.5.2.
It seems that `ContextVar` cannot be dumped for multi-processing.

```
Traceback (most recent call last):
  File "xxxxxxxxxx/dbt/.venv/bin/sqlfluff", line 8, in <module>
    sys.exit(cli())
  File "xxxxxxxxxx/dbt/.venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "xxxxxxxxxx/dbt/.venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "xxxxxxxxxx/dbt/.venv/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "xxxxxxxxxx/dbt/.venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "xxxxxxxxxx/dbt/.venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "xxxxxxxxxx/dbt/.venv/lib/python3.10/site-packages/sqlfluff/cli/commands.py", line 605, in lint
    result = lnt.lint_paths(
  File "xxxxxxxxxx/dbt/.venv/lib/python3.10/site-packages/sqlfluff/core/linter/linter.py", line 1206, in lint_paths
    for i, linted_file in enumerate(runner.run(expanded_paths, fix), start=1):
  File "xxxxxxxxxx/dbt/.venv/lib/python3.10/site-packages/sqlfluff/core/linter/runner.py", line 142, in run
    for lint_result in self._map(
  File "/opt/homebrew/Cellar/python@3.10/3.10.12/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
  File "/opt/homebrew/Cellar/python@3.10/3.10.12/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/pool.py", line 540, in _handle_tasks
    put(task)
  File "/opt/homebrew/Cellar/python@3.10/3.10.12/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/opt/homebrew/Cellar/python@3.10/3.10.12/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle '_contextvars.ContextVar' object
```

**Comment by nevdelap** [comment]
Hi there. I have the same symptom, but with these versions...

* sqlfluff versions: 2.1.0 and 2.1.1
* dbt-snowflake versions: 1.5.0 and 1.5.1
* sqlfluff-templater-dbt versions: 2.1.0 and 2.1.1

I'm trying to get to the bottom of it, but in the meantime here is more info.

## Steps to repro with the older versions.

With this directory/file structure:

```
$ pwd
/tmp/sqlfluff-issue-4930
$ tree -a
.
├── .dbt
│   └── profiles.yml
├── dbt_project.yml
├── Dockerfile
├── .sqlfluff
└── whatever.sql
```

### profiles.yml

I have the environment variables set.

```
---
version: 2

# Used from the development environment, using dev's usernames and keys.
default:
  target: dev
  outputs:
    dev:
      type: snowflake
      account: bogus.ap-southeast-2

      user: "{{ env_var('DBT_USER') }}"
      private_key_path: "{{ env_var('DBT_KEY_FILE') }}"
      private_key_passphrase: "{{ env_var('DBT_KEY_PASSPHRASE') }}"

      database: BOGUS
      schema: BOGUS
```

### dbt_projects.yml

```
---

name: 'sqlfluffissue4930'
version: '1.0.0'
config-version: 2

profile: 'default'
```

### .sqlfluff

```
[sqlfluff]
templater = dbt

[sqlfluff:templater:dbt]
profile = default
profiles_dir = /.dbt
project_dir = /sql
target = dev
```

### Dockerfile

```
FROM sqlfluff/sqlfluff:2.1.0
USER 0
RUN pip install --no-cache-dir dbt-snowflake==1.5.0 sqlfluff-templater-dbt==2.1.0
USER 5000
```

### whatever.sql

```
SELECT *
FROM wherever
```

### Commands To Run it

```
docker build . -t sqlfluff-issue-4930
docker run \
    --rm \
    -u "$(id -u):$(id -g)" \
    -v "$PWD/.dbt/profiles.yml":/.dbt/profiles.yml \
    -v ".:/sql" \
    -e DBT_USER="$DBT_USER" \
    -e DBT_KEY_FILE=/.dbt/rsa_key.p8 \
    -e DBT_KEY_PASSPHRASE="${DBT_KEY_PASSPHRASE:-}" \
    -w /sql \
    sqlfluff-issue-4930 \
    lint \
    --dialect snowflake \
    /sql/whatever.sql
```

### Output

With the minor differences that @MarcBook is using 3.10 and the image is 3.9 add the directory it is all in, it's the same.

```
=== [dbt templater] Sorting Nodes...
00:56:03  Registered adapter: snowflake=1.5.0
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
Traceback (most recent call last):
  File "/app/.venv/bin/sqlfluff", line 8, in <module>
    sys.exit(cli())
  File "/app/.venv/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/app/.venv/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/app/.venv/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/app/.venv/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/app/.venv/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/cli/commands.py", line 605, in lint
    result = lnt.lint_paths(
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/core/linter/linter.py", line 1206, in lint_paths
    for i, linted_file in enumerate(runner.run(expanded_paths, fix), start=1):
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/core/linter/runner.py", line 106, in run
    for fname, partial in self.iter_partials(fnames, fix=fix):
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/core/linter/runner.py", line 59, in iter_partials
    for fname, rendered in self.iter_rendered(fnames):
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/core/linter/runner.py", line 46, in iter_rendered
    yield fname, self.linter.render_file(fname, self.config)
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/core/linter/linter.py", line 879, in render_file
    return self.render_string(raw_file, fname, config, encoding)
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/core/linter/linter.py", line 850, in render_string
    templated_file, templater_violations = self.templater.process(
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/core/templaters/base.py", line 53, in _wrapped
    return func(self, in_str=in_str, fname=fname, config=config, **kwargs)
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff_templater_dbt/templater.py", line 350, in process
    processed_result = self._unsafe_process(fname_absolute_path, in_str, config)
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff_templater_dbt/templater.py", line 466, in _unsafe_process
    node = self._find_node(fname, config)
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff_templater_dbt/templater.py", line 403, in _find_node
    results = [self.dbt_manifest.expect(uid) for uid in selected]
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff_templater_dbt/templater.py", line 403, in <listcomp>
    results = [self.dbt_manifest.expect(uid) for uid in selected]
  File "/app/.venv/lib/python3.9/site-packages/dbt/graph/selector_methods.py", line 329, in search
    root = Path(cv_project_root.get())
LookupError: <ContextVar name='project_root' at 0x7f77e44c4ef0>
```

**Comment by nicolas-gaillard** [comment]
I have the same issue but I don't understand how the mentioned commit can break a previous release. It worked very well until today, I'm a bit confused...

(`sqlfluff==2.1.1 sqlfluff-templater-dbt==2.1.1 dbt-snowflake==1.5.2 dbt-core==1.5.2`)

**Comment by dbeatty10** [comment]
@nevdelap this portion of your error output looks like the [suspected dbt-core commit](https://github.com/dbt-labs/dbt-core/commit/ca73a2aa15aba760abd4503dab7a16d6023c389f#diff-9aa4de7075574ef36d838665148e5858294fae6fd66c0eed1fcf2033a34ee84eR329) to me:
```
root = Path(cv_project_root.get())
```

I'm _guessing_ that although you are using dbt-snowflake versions 1.5.0 and 1.5.1 that it is still using dbt-core 1.5.2 along with them.

You can check your **Core** version versus your (adapter) **Plugin** version(s) by running this:
```
dbt --version
```

We (dbt Labs) merged https://github.com/dbt-labs/dbt-core/issues/7937 yesterday, and it will be included in the next patch release of dbt-core 1.5x.

Based on the manual testing I did for sqlfluff along with that PR, I'm optimistic that it will eliminate this error message once we do that patch release of dbt-core.

**Comment by nevdelap** [comment]
Hi @dbeatty10. You're right. Although dbt isn't in the container (or isn't on the path) to run `dbt --version` I hadn't noticed in the `docker build` output dbt-core is installed. Adding `--no-deps` to the `pip install` in the `Dockerfile` it doesn't and it gets past the error. :+1: 

I don't know why built sqlfluff docker containers in CI are having the same symptom yet, but do think it's a separately solvable issue.
